File: bin/pulser/pulser.m
## -*- texinfo -*-
## @deftypefn {} {[@var{V}, @var{TS}, @var{FFT}, @var{HAX}] = } pulser(@var{F}, @var{W}, @var{NP}, @var{SAVE}, @var{FS}, @var{OUTFILE})
## Generate, play, and (optionally) save a train of Gaussian pulses.
##
## All arguments are optional; their types, default values, and descriptions are as follows:
##
## @table @asis
## @item @qcode{real F = 1. (kHz)}
## The pulse frequency in kilohertz. 
##
## @item @qcode{real W = 100. (us)}
## The width (i.e., stdev or sigma) of the individual Gaussian pulses in microseconds.
##
## @item @qcode{real integer NP = 1000}
## The total number of pulses in the pulse train.
##
## @item @qcode{bool SAVE = true}
## Whether to save the results (both audio and graphical).
##
## @item @qcode{real FS = 44100 (Hz)}
## The sampling frequency in Hertz. Common values include 22050, 44100,
## and 96000. SUPPORT FOR NON-STANDARD VALUES IS NOT GUARANTEED.
##
## @item @qcode{string OUTFILE = ''}
## The desired name (prefix) of the output files.
## Defaults to 'pulser_' followed by the input parameters.
## 
## @end table
##
## If one output is requested, @var{V} is a vector containing the signal's normalized sample values.
##
## If a second output is requested, @var{TS} is the corresponding time vector in seconds.
##
## If a third output is requested, @var{FFT} is a struct containing the frequency vector
##   @var{FFT.f} and relative normalized amplitudes @var{FFT.A} from the discrete Fourier transform.
##
## If a fourth output is requested, @var{HAX} is a vector containing the axes handles of the two plots.
##
## NOTE: The signal can be looped for longer output:
##
## At the @qcode{Octave} prompt:
## @example
## @group
## for k = 1:10; sound(@var{V}, @var{FS}); done	% loop 10 times
## % or
## while true; sound(@var{V}, @var{FS}); done	% loop indefinitely
## @end group
## @end example
##
## At the system prompt, using @qcode{ffplay} from @qcode{ffmpeg}:
## @example
## @group
## ffplay -loop 10 -loglevel quiet pulser.wav	# loop 10 times
## # or
## ffplay -loop 0 -loglevel quiet pulser.wav	# loop indefinitely
## @end group
## @end example
##
## ~ Copyright Mark J. Duvall ~ mjduvall at hawaii.edu ~ 03/2024 ~
##
## @end deftypefn

function [ V TS FFT HAX ] = pulser( F = 1., W = 100., NP = 1000, SAVE = true, FS = 44100., OUTFILE = '' )

%Copyright (C) 2024 Mark J. Duvall / T. Rocks Science
%
%    This program is free software: you can redistribute it and/or modify
%    it under the terms of the GNU General Public License as published by
%    the Free Software Foundation, either version 3 of the License, or
%    (at your option) any later version.
%
%    This program is distributed in the hope that it will be useful,
%    but WITHOUT ANY WARRANTY; without even the implied warranty of
%    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
%    GNU General Public License for more details.
%
%    You should have received a copy of the GNU General Public License
%    along with this program.  If not, see <http://www.gnu.org/licenses/>.

% init
% parameter checks
mustBeGreaterThan(1/(1000.*F), W*1.e-6)
mustBeLessThan(F, FS/2.)
mustBeInteger(NP)
mustBeInteger(FS)
% output filename prefix
if isempty(OUTFILE)
  Fstr = sprintf('%.3f', F);
  Wstr = sprintf('%d', round(W));
  FSstr = sprintf('%.1f', FS/1000.);
  OUTFILE = sprintf('pulser_%skHz-%sus-%skHz', Fstr, Wstr, FSstr);
endif
% basic quantities
T = 1/(1000.*F);						% period between pulses, in seconds
NSamples_pulse = round(T*FS);					% number of samples per pulse
NSamples_total = NSamples_pulse * NP;				% total number of samples in complete pulse train
F_NY = floor(FS/2.);						% Nyquist frequency (half the sampling rate)

% main -- single pulse
t = linspace(0., T, NSamples_pulse);				% time vector for single pulse
L = length(t);							% for convenience; should equal NSamples_pulse
mu = T/2;							% simplify parameters for Gaussian
sigma = W*1.e-6;						% simplify parameters for Gaussian
v = 1/(sigma*sqrt(2*pi)) * exp( -(t-mu).^2/(2*sigma^2) );	% generate Gaussian pulse
v = v/max(v) * (0.5)^2;						% normalize and reduce volume to 50%

% main -- pulse train
TS = linspace(0, NP*T, NP*NSamples_pulse);			% time vector for complete signal
V = zeros(1, NP*L);						% allocate signal vector
for k = 0:(NP-1)						% build complete signal by duplicating the single pulse
  V(1+k*L:(k+1)*L) = v;
end

% calculate spectrum
FFT.nPulses = min([10 NP]);					% number of pulses to analyze per DFT window (10 or the total number of pulses if less than 10)
FFT.nfft = L*FFT.nPulses;					% set window length for DFT algorithm; this is usually a power of 2, but in this case we know the cycle length of the signal, so we can match it exactly
FFT.A = fft(V, FFT.nfft);					% perform the Fourier transform
FFT.L = floor(length(FFT.A)/2);					% get the half-length of the transform result (see below)
FFT.A = abs(FFT.A(1:FFT.L));					% only the absolute value of the first half of the transform result is physically meaningful
FFT.A = FFT.A / max(FFT.A);					% normalize
FFT.f = linspace(0, F_NY, FFT.L);				% frequency vector for plot, in Hz

% plots
% signal
NPulses_to_show = min([10 NP]);
f1 = figure;
HAX(1) = subplot(2, 1, 1);
HAX(2) = subplot(2, 1, 2);
p(1) = plot(HAX(1), t*1.e3, v);
p(2) = plot(HAX(2), TS(1:NSamples_pulse*NPulses_to_show)*1.e3, V(1:NSamples_pulse*NPulses_to_show));
xla(1) = xlabel(HAX(1), 't (ms), Single Pulse');
if NP <= 10
  xla2str = sprintf('t (ms), All %d Pulses', NPulses_to_show);
else
  xla2str = sprintf('t (ms), First %d Pulses', NPulses_to_show);
endif
xla(2) = xlabel(HAX(2), xla2str);
TITLE = title(HAX(1), 'Output Signal');
% spectrum
f2 = figure();
HAX(3) = axes;
p(3) = plot(HAX(3), FFT.f/1000., FFT.A, 'linewidth', 3);
xla(3) = xlabel(HAX(3), 'Frequency (kHz)');
yla(3) = ylabel(HAX(3), 'Normalized Relative Amplitude');
grid on
Fmax = min([ FFT.f(end)/1000. F*10 ]);
xlim([0 Fmax])
% show figures
drawnow

% play
sound(V, FS);

% save if desired
if SAVE
  printf('\n')
  disp 'Writing signal to .wav and writing plots to .svg/.png...'
  audiowrite(sprintf('%s.wav', OUTFILE), V, FS)
  set(HAX, 'fontsize', 10)
  print(f1, sprintf('%s.svg',OUTFILE))
  print(f1, sprintf('%s.png',OUTFILE))
  print(f1, sprintf('%s_spectrum.svg',OUTFILE))
  print(f1, sprintf('%s_spectrum.png',OUTFILE))
  disp 'Done.'
  printf('\n')
endif

% all pau!   )
endfunction

File: bin/pulser/pulser.py
"""
PULSE GENERATOR CODE (PYTHON3) - pulse.py

  This code is equivalent to Mark's pulse generator
  code but this version is written and run in python3.

  Jeffrey Gabriel Yepez
  Dated: 19 March 2024

  Minor modifications by Mark J. Duvall
  Dated: 22 March 2024
"""

import pyaudio
import time
import numpy as np
import matplotlib.pyplot as plt
import wavio

########################################################
# INITIALIZATION

PLOT_SIGNAL = True
SAVE_SIGNAL = False
PLAY_SIGNAL = True

SAVE_WAV_FILE = True

PLOT_SPECTRUM = True
SAVE_SPECTRUM = False

fs = 44100  # Sampling rate (Hz)

n_pulses = 1000
pulse_dur = 0.001
duration = n_pulses * pulse_dur  # In seconds

a = 0.25 # Amplitude in range [0.0, 1.0]

N = int(duration * fs) # Number of samples

########################################################
# STEP 1: TIME SIGNAL

tstep = 1 / fs
t = np.linspace(0, (N - 1) * tstep, N)

# Generate samples for gaussian pulse.
mu0 = 0.5 # Position of first gaussian pulse in time domain
sigma = 0.0001 # Spread of gaussian
y = 0 # Initialize waveform variable
for n in range(n_pulses):
    # Note that the pulse train is created by changing mu and adding the signal spaces together, where "mu = duration * (n + l) / n_pulses" and n is the iterator.
    y += a * np.exp(-((t - (duration * (n + mu0) / n_pulses)) / sigma) ** 2).astype(np.float32)
    
########################################################
# STEP 2: AMPLITUDE SPECTRUM

fstep = fs / N
f = np.linspace(0, (N - 1) * fstep, N)

# Compute the Fourier transform with the DC component (i.e., signal mean) removed
# Note: We do not subtract the mean from the signal itself
#   (e.g., by executing 'y=y-np.mean(y)' because we want
#   to keep the signal polarity uniform. Pulses that are exclusively
#   positive or negative are more useful in lab settings.
Y = np.abs(np.fft.fft(y-np.mean(y))) / N
# Normalize:
Y = Y / np.max(Y)

f_plot = f[0:int(N/2 + 1)]
Y_plot = Y[0:int(N/2 + 1)]

########################################################
# STEP 3: PLOTTING

# plot up to 10 pulses
N_pulse = np.floor(pulse_dur * fs)
N_plot = int(np.minimum(10, n_pulses) * N_pulse)
plt.plot(t[0:N_plot], y[0:N_plot])
plt.xlabel("Time (s)")
plt.title("Pulse Signal")
if SAVE_SIGNAL:
    plt.savefig("signal.pdf")
if PLOT_SIGNAL:
    plt.show()
else:
    plt.close()

pulse_freq = 1.e-3 / pulse_dur
plt.plot(f_plot/1000, Y_plot, linewidth=2)
plt.xlim(0, np.minimum(22.050, 7.5*pulse_freq))
plt.grid()
plt.xlabel("Frequency (kHz)")
plt.ylabel("Normalized Relative Amplitude")
plt.title("Spectrum")
if SAVE_SPECTRUM:
    plt.savefig("spectrum.pdf")
if PLOT_SPECTRUM:
    plt.show()
else:
    plt.close()

########################################################
# STEP 4: SIGNAL OUTPUT

if PLAY_SIGNAL:
    p = pyaudio.PyAudio()

    # Convert signal to bytes to play signal.
    output_bytes = (y).tobytes()

    # For paFloat32 sample values must be in range [-1.0, 1.0].
    stream = p.open(format=pyaudio.paFloat32,
                channels=1,
                rate=fs,
                output=True)
                
    # Plays waveform. May repeat with different volume values.
    start_time = time.time()
    stream.write(output_bytes)
    print("Played sound for {:.4f} seconds".format(time.time() - start_time))

    stream.stop_stream()
    stream.close()

########################################################
# STEP 5: SAVE SIGNAL

if SAVE_WAV_FILE:
    print('Saving .wav output...')
    # Ensure that highest value is in 16-bit range
    audio = y * (2**15 - 1) / np.max(np.abs(y))
    # Convert to 16-bit data
    audio = audio.astype(np.int16)
    # write out signal
    wavio.write('signal.wav', y, fs, sampwidth=2)
    print('Done.')

########################################################
# All pau!   )

print("Finished. Exiting...")
# exit()
